dm/master/server.go

```diff
@@ -895,43 +925,79 @@ func (s *Server) getStatusFromWorkers(ctx context.Context, sources []string, tas
 				Source: source,
 			},
 		}
-		workerRespCh <- resp
-		return false
+		if worker != "" {
+			resp.SourceStatus.Worker = worker
+		}
+		setWorkerResp(resp)
 	}
 
 	var wg sync.WaitGroup
 	for _, source := range sources {
-		wg.Add(1)
-		go s.ap.Emit(ctx, 0, func(args ...interface{}) {
-			defer wg.Done()
-			sourceID, _ := args[0].(string)
-			worker := s.scheduler.GetWorkerBySource(sourceID)
-			if worker == nil {
-				err := terror.ErrMasterWorkerArgsExtractor.Generatef("%s relevant worker-client not found", sourceID)
-				handleErr(err, sourceID)
-				return
+		var (
+			workers       []*scheduler.Worker
+			workerNameSet = make(map[string]struct{})
+			err2          error
+		)
+		if relayWorker {
+			workers, err2 = s.scheduler.GetRelayWorkers(source)
+			if err2 != nil {
+				handleErr(err2, source, "")
+				continue
 			}
-			resp, err := worker.SendRequest(ctx, workerReq, s.cfg.RPCTimeout)
-			var workerStatus *pb.QueryStatusResponse
-			if err != nil {
-				workerStatus = &pb.QueryStatusResponse{
-					Result:       false,
-					Msg:          err.Error(),
-					SourceStatus: &pb.SourceStatus{},
-				}
-			} else {
-				workerStatus = resp.QueryStatus
+			// returned workers is not duplicated
+			for _, w := range workers {
+				workerNameSet[w.BaseInfo().Name] = struct{}{}
 			}
-			workerStatus.SourceStatus.Source = sourceID
-			workerRespCh <- workerStatus
-		}, func(args ...interface{}) {
-			defer wg.Done()
-			sourceID, _ := args[0].(string)
-			handleErr(terror.ErrMasterNoEmitToken.Generate(sourceID), sourceID)
-		}, source)
+		}
+
+		// subtask workers may have been found in relay workers
+		taskWorker := s.scheduler.GetWorkerBySource(source)
+		if taskWorker != nil {
+			if _, ok := workerNameSet[taskWorker.BaseInfo().Name]; !ok {
+				workers = append(workers, taskWorker)
+			}
+		}
+
+		if len(workers) == 0 {
+			err := terror.ErrMasterWorkerArgsExtractor.Generatef("%s relevant worker-client not found", source)
+			handleErr(err, source, "")
+			continue
+		}
+
+		for _, worker := range workers {
```

Comment:
Will we get repeat status from worker here?

Reply:
we use `workerNameSet` to avoid repeating

Reply:
Since every worker only bound to one source, It LGTM now.